benchmarks/nel/scripts/candidate_generation/embeddings.py

```diff
@@ -1,32 +1,46 @@
-""" Candidate generation. """
-from pathlib import Path
-from typing import Iterator, Dict
+""" Candidate generation via distance in embedding space. """
+from typing import Iterable
+
+import numpy
+from sklearn.neighbors import NearestNeighbors
 
-import spacy
-from spacy import Language
 from spacy.kb import KnowledgeBase, Candidate
 from spacy.tokens import Span
-
-from scripts.datasets.dataset import Dataset
-
-
-_pipelines: Dict[Path, Language] = {}
-
-
-def create_candidates(dataset_name: str, kb: KnowledgeBase, span: Span) -> Iterator[Candidate]:
-    """ Identifies entity candidates via their embeddings.
-    kb (KnowledgeBase): KnowledgeBase containing all possible entity candidates.
-    span (Span): Span to match potential entity candidates with.
-    """
-
-    # todo @RM
-    #  - add temp getters to KB class to access all information necessary to create candidate objects
-    #  - infer vector for span.text
-    #  - compute distances between KB entity vectors and span.text vector
-    #  - define offcut/top-n in config, return relevant candidates
-    #  - evaluate results, compare with get_candidates()
-    # nlp_path = Dataset.assemble_paths(dataset_id)["nlp_base"]
-    # if nlp_path not in _pipelines:
-    #     _pipelines[nlp_path] = spacy.load(nlp_path)
-
-    return kb.get_alias_candidates(span.text)
+from .base import NearestNeighborCandidateSelector
+import fuzzyset
+
+
+class CandidateSelector(NearestNeighborCandidateSelector):
+    """ Callable object selecting candidates as nearest neighbours in embedding space. """
+
+    def _init_lookup_structure(self, kb: KnowledgeBase, max_n_candidates: int, **kwargs) -> NearestNeighbors:
+        container = NearestNeighbors(n_neighbors=max_n_candidates, metric="cosine")
+        container.fit(numpy.asarray([kb.get_vector(ent_id) for ent_id in kb.get_entity_strings()]))
+
+        return container
+
+    def _fetch_candidates(
+        self, dataset_id: str, span: Span, kb: KnowledgeBase, max_n_candidates: int, lexical_similarity_cutoff: float = .5
+    ) -> Iterable[int]:
+        target_vec = self._pipelines[dataset_id](span.text).vector
+        if not isinstance(target_vec, numpy.ndarray):
+            target_vec = target_vec.get()
+
+        nn_idx = self._lookup_structs[dataset_id].kneighbors(target_vec.reshape((1, -1)))[1][0]
+        ent_ids = kb.get_entity_strings()
+        nn_entities = {ent_ids[i]: self._entities[dataset_id][ent_ids[i]] for i in nn_idx}
+        candidate_entity_ids = {
+            match[0] for match in
+            [
+                (nne, fuzzyset._distance(name.lower(), span.text.lower()))
```

Comment:
Might make sense to call the underlying distance function directly rather than proxying through fuzzyset, though I guess that expands the number of APIs we rely on.